File: gilgamesh/snes/registers.py
from copy import copy
from typing import Dict, Optional

from gilgamesh.snes.state import State


class Register:
    def __init__(self, state: State, accumulator: bool):
        self.state = state
        self.accumulator = accumulator
        self.lo: Optional[int] = None
        self.hi: Optional[int] = None

    @property
    def size(self) -> int:
        return self.state.a_size if self.accumulator else self.state.x_size

    def copy(self, state: State) -> "Register":
        register = copy(self)
        register.state = state
        return register

    def get(self) -> Optional[int]:
        if self.size == 1:
            return self.lo
        else:
            if (self.lo is None) or (self.hi is None):
                return None
            return (self.hi << 8) | self.lo

    def set(self, value: Optional[int]) -> None:
        if value is None:
            if self.size == 1:
                self.lo = None
            else:
                self.lo = self.hi = None
        elif self.size == 1:
            self.lo = value & 0xFF
        else:
            self.lo = value & 0xFF
            self.hi = (value >> 8) & 0xFF

    def get_whole(self) -> Optional[int]:
        if (self.lo is None) or (self.hi is None):
            return None
        return (self.hi << 8) | self.lo

    def set_whole(self, value: Optional[int]) -> None:
        if value is None:
            self.lo = self.hi = None
        else:
            self.lo = value & 0xFF
            self.hi = (value >> 8) & 0xFF


class Registers:
    def __init__(self, state: State):
        self.state = state
        self.a = Register(state, True)
        self.x = Register(state, False)
        self.y = Register(state, False)

    def copy(self, state: State) -> "Registers":
        registers = copy(self)
        registers.state = state
        registers.a = registers.a.copy(state)
        registers.x = registers.x.copy(state)
        registers.y = registers.y.copy(state)
        return registers

    def snapshot(self) -> Dict[str, Optional[int]]:
        return {
            "a": self.a.get(),
            "x": self.x.get(),
            "y": self.y.get(),
        }

File: src/snes/registers.rs
use crate::snes::state::StateRegister;

#[derive(Copy, Clone)]
struct Register {
    state: StateRegister,
    is_accumulator: bool,
    lo: Option<u16>,
    hi: Option<u16>,
}

impl Register {
    pub fn new(state: StateRegister, is_accumulator: bool) -> Self {
        Self {
            state,
            is_accumulator,
            lo: None,
            hi: None,
        }
    }

    pub fn size(&self) -> usize {
        if self.is_accumulator {
            self.state.a_size()
        } else {
            self.state.x_size()
        }
    }

    pub fn get(&self) -> Option<u16> {
        match self.size() {
            1 => self.lo,
            _ => self.get_whole(),
        }
    }

    pub fn get_whole(&self) -> Option<u16> {
        if self.lo.is_none() || self.hi.is_none() {
            None
        } else {
            let (hi, lo) = (self.hi.unwrap(), self.lo.unwrap());
            Some((hi << 8) | lo)
        }
    }

    pub fn set(&mut self, value: Option<u16>) {
        match value {
            Some(v) => {
                self.lo = Some(v & 0xFF);
                if self.size() > 1 {
                    self.hi = Some((v >> 8) & 0xFF);
                }
            }
            None => match self.size() {
                1 => self.lo = None,
                _ => {
                    self.lo = None;
                    self.hi = None;
                }
            },
        }
    }

    pub fn set_whole(&mut self, value: Option<u16>) {
        match value {
            Some(v) => {
                self.lo = Some(v & 0xFF);
                self.hi = Some((v >> 8) & 0xFF);
            }
            None => {
                self.lo = None;
                self.hi = None;
            }
        }
    }
}

#[derive(Copy, Clone)]
pub struct Registers {
    state: StateRegister,
    a: Register,
    x: Register,
    y: Register,
}

impl Registers {
    pub fn new(state: StateRegister) -> Self {
        Self {
            state,
            a: Register::new(state, true),
            x: Register::new(state, false),
            y: Register::new(state, false),
        }
    }
}
